sekoia_automation/connector/event_aggregator.py

```diff
@@ -0,0 +1,187 @@
+        self.f_must_stop = True
+
+    def run(self) -> None:
+        try:
+            while not self.f_must_stop:
+                time.sleep(self.delay)
+                for aggregated_event in self.event_aggregator.flush_all_ttl():
+                    self.on_flush_func(aggregated_event)
+        finally:
+            for aggregated_event in self.event_aggregator.flush_all():
+                self.on_flush_func(aggregated_event)
+
+
+class EventAggregator:
+    ttl_thread: EventAggregatorTTLThread | None
+    lock: threading.Lock
+
+    def __init__(self, aggregation_definitions: dict[str, list[Fingerprint]]):
+        self.aggregation_definitions = aggregation_definitions
+        self.aggregations: dict[str, Aggregation] = dict()
+        self.lock = threading.Lock()
+
+    def start_flush_on_ttl(
+        self, on_flush_func: Callable[[dict], None], delay: float = 10
+    ):
+        """
+        This method MUST be called to trigger the execution of the ttl thread
+        """
+        self.ttl_thread = EventAggregatorTTLThread(
+            event_aggregator=self, on_flush_func=on_flush_func, delay=delay
+        )
+        self.ttl_thread.start()
+
+    def stop(self):
+        """
+        This method is only meant to stop the ttl thread
+        """
+        if self.ttl_thread:
+            self.ttl_thread.stop()
+            self.ttl_thread.join()
+
+    def flush_all(self) -> list[dict]:
+        """
+        returns all the aggregated events and remove them from the ongoing aggregations
+        """
+        aggregated_events = []
+
+        # we prevent any actions on the aggregations while flushing
+        with self.lock:
+            for aggregation in self.aggregations.values():
+                if aggregation.count > 0:
+                    aggregated_events.append(aggregation.get_aggregated_event())
+
+            self.aggregations = dict()
+        return aggregated_events
+
+    def flush_all_ttl(self) -> list[dict]:
+        """
+        Returns (and delete) all the events we aggregate for at least the ttl time
+        """
+        aggregated_events = []
+        # we prevent any actions on the aggregations while flushing
+        with self.lock:
+            for event_hash in list(self.aggregations.keys()):
+                aggregation = self.aggregations[event_hash]
+                if (
+                    aggregation.start + datetime.timedelta(seconds=aggregation.ttl)
+                    < datetime.datetime.utcnow()
+                ):
+                    if aggregation.count > 0:
+                        aggregated_events.append(aggregation.get_aggregated_event())
+
+                    del self.aggregations[event_hash]
+
+        return aggregated_events
+
+    def get_hash(self, event: dict) -> Tuple[str, int] | None:
+        """
+        Returns the hash to fingerprint the specified event and its ttl
+        """
+        event_dialect_uuid = event["sekoiaio"]["intake"]["dialect_uuid"]
```

Comment:
I'm not sure this attribute is set in the events. The workflow sets it, but in the connectors we usually only have the intake key.

Reply:
I removed the need to explicit the dialect as an aggregation engine will be configured for each connector with their specific dialects